File: PyPI/src/spreen_wiki/configuration.py
import os
import re
from importlib.resources import files
from typing import Any

import yaml

CONFIG_FILENAME = '.spreen.yml'
DEFAULT_EXCLUDED_DIRS = ['github-wiki-organisers', 'wikis-by-owner']

# Built-in labels: how the first line of a wiki page is parsed (regexp),
# which namespace collects pages without a declaration (no_declaration),
# which namespaces the count report covers (unknown_namespaces) and which
# namespace the LLM export targets (llm_target_namespace). A config file
# can override any of these, add languages or add group_by criteria.
DEFAULT_LABELS: dict[str, Any] = {
    'Owner': {
        'regexp': r'[Oo]wner:\s?',
        'languages': {
            'English': {
                'no_declaration': 'Unowned',
                'unknown_namespaces': [
                    'Unknown Owner nor Necessity',
                    'Unowned but Necessary',
                    'Unowned',
                ],
                'llm_target_namespace': 'Unknown Owner nor Necessity',
            },
            'Japanese': {
                'no_declaration': 'Owner記名なし',
                'unknown_namespaces': [
                    'Ownerチームが不明だが必要なページ群',
                    'Ownerチーム・要or不要が不明なページ群',
                    'Owner記名なし',
                ],
                'llm_target_namespace': 'Ownerチーム・要or不要が不明なページ群',
            },
        },
    },
    'Category': {
        'regexp': r'[Cc]ategory:\s?',
        'languages': {
            'English': {
                'no_declaration': 'Uncategorised',
                'unknown_namespaces': ['Uncategorised'],
                'llm_target_namespace': 'Uncategorised',
            },
            'Japanese': {
                'no_declaration': 'Category記載なし',
                'unknown_namespaces': ['Category記載なし'],
                'llm_target_namespace': 'Category記載なし',
            },
        },
    },
}


def default_template_dir() -> str:
    """The `templates/` directory shipped with this package."""
    return str(files(__package__).joinpath('templates'))


def _presence(value: str | None) -> str | None:
    """Treats empty strings the same as None so that callers can pass
    values straight from optional environment variables or CLI flags."""
    return value if value else None


def _deep_merge(base: dict[str, Any], override: dict[str, Any]) -> dict[str, Any]:
    merged = dict(base)
    for key, new_value in override.items():
        old_value = merged.get(key)
        if isinstance(old_value, dict) and isinstance(new_value, dict):
            merged[key] = _deep_merge(old_value, new_value)
        else:
            merged[key] = new_value
    return merged


class Configuration:
    """Resolves the runtime configuration for a wiki checkout. Every value
    is looked up with the same precedence: explicit keyword argument, then
    the `.spreen.yml` file under base_path (or config_path), then
    the ORGANISATION_NAME environment variable (organisation only), then
    the built-in defaults."""

    def __init__(self, base_path: str, config_path: str | None = None,
                 organisation: str | None = None,
                 repository: str | None = None,
                 wiki_url: str | None = None,
                 owner_base_url: str | None = None,
                 excluded_dirs: list[str] | None = None,
                 template_dir: str | None = None) -> None:
        self._file = self._load_file(
            config_path or os.path.join(base_path, CONFIG_FILENAME))
        self.organisation = self._resolve_organisation(organisation)
        self.repository = self._resolve(repository, 'repository')
        self.wiki_url = (self._resolve(wiki_url, 'wiki_url')
                         or self._default_wiki_url())
        self.owner_base_url = (self._resolve(owner_base_url, 'owner_base_url')
                               or self._default_owner_base_url())
        self.excluded_dirs: list[str] = (
            excluded_dirs if excluded_dirs is not None
            else self._file.get('exclude', DEFAULT_EXCLUDED_DIRS))
        self.template_dir: str = (
            template_dir or self._file.get(
                'template_dir', default_template_dir()))
        self._labels = _deep_merge(
            DEFAULT_LABELS, self._file.get('labels', {}))

    def group_bys(self) -> list[str]:
        return list(self._labels.keys())

    def languages(self, group_by: str) -> list[str]:
        return list(self._label_for(group_by).get('languages', {}).keys())

    def target_regexp(self, group_by: str) -> re.Pattern[str]:
        return re.compile(self._label_for(group_by).get('regexp', ''))

    def no_declaration(self, group_by: str, language: str) -> str:
        value: str = self._language_labels(group_by, language).get(
            'no_declaration', '')
        return value

    def unknown_namespaces(self, group_by: str,
                           language: str) -> list[str]:
        value: list[str] = self._language_labels(group_by, language).get(
            'unknown_namespaces', [])
        return value

    def llm_target_namespace(self, group_by: str, language: str) -> str:
        value: str = self._language_labels(group_by, language).get(
            'llm_target_namespace', '')
        return value

    def path_to_template(self, group_by: str, language: str) -> str:
        return os.path.join(self.template_dir, group_by.lower(),
                            f'{language.lower()}.md')

    # private

    @staticmethod
    def _load_file(path: str) -> dict[str, Any]:
        if not os.path.isfile(path):
            return {}
        with open(path, encoding='utf-8') as f:
            return yaml.safe_load(f) or {}

    def _resolve(self, explicit: str | None, key: str) -> str | None:
        return _presence(explicit) or _presence(self._file.get(key))

    def _resolve_organisation(self, explicit: str | None) -> str | None:
        return (self._resolve(explicit, 'organisation')
                or _presence(os.environ.get('ORGANISATION_NAME')))

    def _default_wiki_url(self) -> str | None:
        if self.organisation and self.repository:
            return (f'https://github.com/{self.organisation}'
                    f'/{self.repository}/wiki')
        return None

    def _default_owner_base_url(self) -> str | None:
        if self.organisation:
            return f'https://github.com/orgs/{self.organisation}/teams/'
        return None

    def _label_for(self, group_by: str) -> dict[str, Any]:
        label: dict[str, Any] = self._labels.get(group_by, {})
        return label

    def _language_labels(self, group_by: str,
                         language: str) -> dict[str, Any]:
        labels: dict[str, Any] = self._label_for(group_by).get(
            'languages', {}).get(language, {})
        return labels

File: RubyGem/lib/spreen/wiki/configuration.rb
# frozen_string_literal: true
# rbs_inline: enabled

require 'yaml'

module Spreen
  module Wiki
    # Resolves the runtime configuration for a wiki checkout. Every value is
    # looked up with the same precedence: explicit keyword argument, then the
    # `.spreen.yml` file under base_path (or config_path), then the
    # ORGANISATION_NAME environment variable (organisation only), then the
    # built-in defaults.
    class Configuration
      CONFIG_FILENAME = '.spreen.yml'
      EMPTY_HASH = {}.freeze #: Hash[String, untyped]
      DEFAULT_TEMPLATE_DIR = File.expand_path(File.join(__dir__.to_s, 'templates'))
      DEFAULT_EXCLUDED_DIRS = %w[spreen-wiki wikis-by-owner].freeze

      # Built-in labels: how the first line of a wiki page is parsed (regexp),
      # which namespace collects pages without a declaration (no_declaration),
      # which namespaces the count report covers (unknown_namespaces) and which
      # namespace the LLM export targets (llm_target_namespace). A config file
      # can override any of these, add languages or add group_by criteria.
      DEFAULT_LABELS = {
        'Owner' => {
          'regexp' => '[Oo]wner:\s?',
          'languages' => {
            'English' => {
              'no_declaration' => 'Unowned',
              'unknown_namespaces' => ['Unknown Owner nor Necessity', 'Unowned but Necessary', 'Unowned'],
              'llm_target_namespace' => 'Unknown Owner nor Necessity'
            },
            'Japanese' => {
              'no_declaration' => 'Owner記名なし',
              'unknown_namespaces' => [
                'Ownerチームが不明だが必要なページ群',
                'Ownerチーム・要or不要が不明なページ群',
                'Owner記名なし'
              ],
              'llm_target_namespace' => 'Ownerチーム・要or不要が不明なページ群'
            }
          }
        },
        'Category' => {
          'regexp' => '[Cc]ategory:\s?',
          'languages' => {
            'English' => {
              'no_declaration' => 'Uncategorised',
              'unknown_namespaces' => ['Uncategorised'],
              'llm_target_namespace' => 'Uncategorised'
            },
            'Japanese' => {
              'no_declaration' => 'Category記載なし',
              'unknown_namespaces' => ['Category記載なし'],
              'llm_target_namespace' => 'Category記載なし'
            }
          }
        }
      }.freeze

      attr_reader :organisation, :repository, :wiki_url, :owner_base_url, :excluded_dirs, :template_dir

      # @rbs!
      #   @organisation: String?
      #   @repository: String?
      #   @wiki_url: String?
      #   @owner_base_url: String?
      #   @excluded_dirs: Array[String]
      #   @template_dir: String
      #   @file: Hash[String, untyped]
      #   @labels: Hash[String, untyped]

      # @rbs base_path: String
      # @rbs config_path: String?
      # @rbs organisation: String?
      # @rbs repository: String?
      # @rbs wiki_url: String?
      # @rbs owner_base_url: String?
      # @rbs excluded_dirs: Array[String]?
      # @rbs template_dir: String?
      # @rbs return: void
      def initialize(base_path:, config_path: nil, organisation: nil, repository: nil, wiki_url: nil,
                     owner_base_url: nil, excluded_dirs: nil, template_dir: nil)
        @file           = load_file(config_path || File.join(base_path, CONFIG_FILENAME))
        @organisation   = resolve_organisation(organisation)
        @repository     = resolve(repository, 'repository')
        @wiki_url       = resolve_wiki_url(wiki_url)
        @owner_base_url = resolve_owner_base_url(owner_base_url)
        @excluded_dirs  = excluded_dirs || file.fetch('exclude', DEFAULT_EXCLUDED_DIRS)
        @template_dir   = template_dir || file.fetch('template_dir', DEFAULT_TEMPLATE_DIR)
        @labels         = deep_merge(DEFAULT_LABELS, file.fetch('labels', EMPTY_HASH))
      end

      # @rbs return: Array[String]
      def group_bys
        labels.keys
      end

      # @rbs group_by: String
      # @rbs return: Array[String]
      def languages(group_by)
        label_for(group_by).fetch('languages', EMPTY_HASH).keys
      end

      # @rbs group_by: String
      # @rbs return: Regexp
      def target_regexp(group_by)
        Regexp.new(label_for(group_by).fetch('regexp', ''))
      end

      # @rbs group_by: String
      # @rbs language: String
      # @rbs return: String
      def no_declaration(group_by, language)
        language_labels(group_by, language).fetch('no_declaration', '')
      end

      # @rbs group_by: String
      # @rbs language: String
      # @rbs return: Array[String]
      def unknown_namespaces(group_by, language)
        language_labels(group_by, language).fetch('unknown_namespaces', [])
      end

      # @rbs group_by: String
      # @rbs language: String
      # @rbs return: String
      def llm_target_namespace(group_by, language)
        language_labels(group_by, language).fetch('llm_target_namespace', '')
      end

      # @rbs group_by: String
      # @rbs language: String
      # @rbs return: String
      def path_to_template(group_by, language)
        File.join(template_dir, group_by.downcase, "#{language.downcase}.md")
      end

      private

      attr_reader :file, :labels

      # Treats empty strings the same as nil so that callers can pass values
      # straight from optional environment variables or CLI flags.
      # @rbs value: String?
      # @rbs return: String?
      def presence(value)
        value unless value.nil? || value.empty?
      end

      # @rbs explicit: String?
      # @rbs key: String
      # @rbs return: String?
      def resolve(explicit, key)
        presence(explicit) || presence(file[key])
      end

      # @rbs explicit: String?
      # @rbs return: String?
      def resolve_organisation(explicit)
        resolve(explicit, 'organisation') || presence(ENV.fetch('ORGANISATION_NAME', nil))
      end

      # @rbs explicit: String?
      # @rbs return: String?
      def resolve_wiki_url(explicit)
        resolve(explicit, 'wiki_url') || default_wiki_url
      end

      # @rbs explicit: String?
      # @rbs return: String?
      def resolve_owner_base_url(explicit)
        resolve(explicit, 'owner_base_url') || default_owner_base_url
      end

      # @rbs path: String
      # @rbs return: Hash[String, untyped]
      def load_file(path)
        return {} unless File.exist?(path)

        YAML.safe_load_file(path) || {}
      end

      # @rbs group_by: String
      # @rbs return: Hash[String, untyped]
      def label_for(group_by)
        labels.fetch(group_by, EMPTY_HASH)
      end

      # @rbs group_by: String
      # @rbs language: String
      # @rbs return: Hash[String, untyped]
      def language_labels(group_by, language)
        label_for(group_by).fetch('languages', EMPTY_HASH).fetch(language, EMPTY_HASH)
      end

      # @rbs return: String?
      def default_wiki_url
        "https://github.com/#{organisation}/#{repository}/wiki" if organisation && repository
      end

      # @rbs return: String?
      def default_owner_base_url
        "https://github.com/orgs/#{organisation}/teams/" if organisation
      end

      # @rbs base: Hash[String, untyped]
      # @rbs override: Hash[String, untyped]
      # @rbs return: Hash[String, untyped]
      def deep_merge(base, override)
        base.merge(override) do |_key, old_value, new_value|
          old_value.is_a?(Hash) && new_value.is_a?(Hash) ? deep_merge(old_value, new_value) : new_value
        end
      end
    end
  end
end
